pontoon/static/js/utils/data.js

```diff
@@ -0,0 +1,90 @@
+
+import React from 'react';
+
+import {ajax} from 'utils/ajax';
+
+
+export class DataManager {
+
+    constructor (component, data) {
+        this.component = component;
+        this.handleResponse = this.handleResponse.bind(this);
+        this.handleSubmit = this.handleSubmit.bind(this);
+        this.refreshData = this.refreshData.bind(this);
+        this.initialData = data;
+    }
+
+    get data () {
+        const {data} = this.component.state;
+        return data;
+    }
+
+    get errors () {
+        const {errors} = this.component.state;
+        return errors;
+    }
+
+    get api () {
+        return this.component.props.api;
+    }
+
+    get requestMethod () {
+        return this.component.props.requestMethod || 'get';
+    }
+
+    get submitMethod () {
+        return this.component.props.submitMethod || this.requestMethod;
+    }
+
+    async refreshData (params) {
+        if (this.initialData) {
+            this.component.setState({data: this.initialData, errors: []});
+            this.initialData = undefined;
+            return
+        }
+        return this.handleResponse(
+            await ajax[this.requestMethod](this.api, params));
+    }
+
+    async handleResponse (response) {
+        const {status} = response;
+        const json = await response.json();
+        if (status === 200) {
+            const {data} = json;
+            this.component.setState({data, errors: {}});
+            return;
+        }
+        const {errors} = json;
+        this.component.setState({errors});
+    }
+
+    async handleSubmit (params) {
+        return this.handleResponse(
+            await ajax[this.submitMethod](this.api, params));
+    }
+}
```

Comment:
I'm confused by the design of this class. It seems to be doing 2 things, holding data and querying that data. I'm quite convinced that in your place I would have written that logic as a function instead of a class, and let a different structure hold the data. 

I guess I need to see it in action in order to understand it better. Could you show me an instance of code where you use it?

Reply:
ultimately it provides a "HOC" function that you can pass in an `object` too - that object can be used to:

- specify the components to be used in rendering
- handle data mangling
- be the (adaptable) "single source of truth"

my reading of the HOC docs is that the data provider can be any object

the slight deviation i made from the standard pattern was that instead of embedding all of the data mangling functionality in the react component its pulled out to a class.

its used in both the admin and project uis

in the admin - v little is overridden in the class - the default could be switched in this case and could use the generic datamanager - although it probs makes sense to keep the default as `get`

the manager is set up here:

https://github.com/mozilla/pontoon/pull/815/files#diff-a58b75723f571a686ce4ade32fc3eeffR61

the refreshData and submit methods are then passed down, and are used here:

https://github.com/mozilla/pontoon/pull/815/files#diff-a58b75723f571a686ce4ade32fc3eeffR13

Using a class allows much more to be overriden - as for example here where the dashboard is specified:

https://github.com/mozilla/pontoon/pull/853/files#diff-c29b1fca17c37a5c3c590455ac4986c0R6

Or the very sparse code required for specifying specific dashboards:

https://github.com/mozilla/pontoon/pull/853/files#diff-917435cc33dad6f9255ad48eb8a951a2R10

https://github.com/mozilla/pontoon/pull/853/files#diff-fb5c63ce89f3847b2c2101ebffeb3092R11

Reply:
updating the code - parts of this functionality could defo be done *inside* the func, but it would still be in a class - ie the `WrappedComponent` - this pattern keeps the composability but also avoids putting all of the functionality into an anonymous class

Reply:
this has been separated into HOC and and data provider for the sake of landing the code